stacker/util.py

```diff
@@ -531,10 +536,69 @@ def create_cache_directories(self):
 
     def get_package_sources(self):
         """Makes remote python packages available for local use."""
+        # Checkout S3 repositories specified in config
+        for config in self.sources.get('s3', []):
+            self.fetch_s3_package(config=config)
         # Checkout git repositories specified in config
         for config in self.sources.get('git', []):
             self.fetch_git_package(config=config)
 
+    def fetch_s3_package(self, config):
+        """Makes a remote S3 archive available for local use
+
+        Args:
+            config (:class:`stacker.config.GitPackageSource`): git config
+                dictionary
+
+        """
+        if config['key'].endswith('.tar.gz'):
+            archive_type = '.tar.gz'
+            dir_name = self.sanitize_uri_path(
+                "s3-%s-%s" % (config['bucket'], config['key'][:-7])
+            )
+        elif config['key'].endswith('.zip'):
+            archive_type = '.zip'
+            dir_name = self.sanitize_uri_path(
+                "s3-%s-%s" % (config['bucket'], config['key'][:-4])
+            )
+        else:
+            raise ValueError(
+                "Archive type could not be determined for S3 object \"%s\" "
+                "in bucket %s." % (config['key'], config['bucket'])
+            )
+
+        # We can skip downloading the archive if it's already been cached
+        cached_dir_path = os.path.join(self.package_cache_dir, dir_name)
+        if not os.path.isdir(cached_dir_path):
+            tmp_dir = tempfile.mkdtemp(prefix='stacker')
+            tmp_package_path = os.path.join(tmp_dir, dir_name)
+            try:
+                tmp_archive_path = os.path.join(tmp_dir,
+                                                dir_name + archive_type)
+                s3_resource = get_session(region=None).resource('s3')
```

Comment:
How do we handle different s3 regions?  Does it make sense for this to use the default stacker bucket if no bucket is specified?  I spent a good chunk of time trying to get this just right in the other uses of s3, so there's probably some useful code there (see the `aws_lambda` hook as well as the methods we use in the provider).